### PR TITLE
Added Ray3d.intersectWithTriangle

### PR DESCRIPTION
**Pull request by saeeedtorabi**
- This new method makes ray-triangle intersection speed 2-3 times faster
- Added few other new APIs
- Added lots of accuracy and performance tests

### CODE REVIEW

File: core/geometry/src/geometry3d/BarycentricTriangle.ts [hunk 1]
```diff
@@ -560,7 +572,7 @@ export class BarycentricTriangle {
   public snapLocationToEdge(
     location: TriangleLocationDetail,
     distanceTolerance: number = Geometry.smallMetricDistance,
```
Comment (saeeedtorabi) [inline core/geometry/src/geometry3d/BarycentricTriangle.ts:574]: I dont remember why I made this chage. I think you told me to but dont remember why!

Comment (dassaf4) [reply core/geometry/src/geometry3d/BarycentricTriangle.ts:574]: It's good. We want that tighter default snap tolerance to correspond to the snap tol in the new method.

**Comment by dassaf4** [comment]
Consider introducing the new method, and contrasting it to similar existing methods, in NextVersion.md. When would someone want to use it over the existing methods, and when not?

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/test/geometry3d/Ray3d.test.ts [hunk 2]
```diff
@@ -282,3 +288,297 @@ describe("Ray3d", () => {
...
+    triangle = BarycentricTriangle.createXYZXYZXYZ(2, 0, 0, 2, 0, 0, 2, 0, 0);
+    intersectionPoint = ray.intersectionWithTriangle(triangle.points[0], triangle.points[1], triangle.points[2]);
+    ck.testUndefined(
+      intersectionPoint,
+      "expect no intersection when we have a degenerate triangle with three equal vertexes"
+    );
+
+    origin = Point3d.create(0, 0, 0);
+    direction = Vector3d.create(1, 1, 0);
+    ray = Ray3d.create(origin, direction);
```
Comment (saeeedtorabi) [inline core/geometry/src/test/geometry3d/Ray3d.test.ts:417]: I need to redefine triangle here:

triangle = BarycentricTriangle.createXYZXYZXYZ(2, 0, 0, 10, 0, 0, 2, 10, 0);

I added it locally and the test passed.

I also added a test for ray intersects inside a triangle and it passed too. Will added them all after you reviewed this.

**Review by dassaf4 — APPROVED** [review]

**Review by ben-polinsky — APPROVED** [review]

**Pull request merged**